Zephyr v3.6.0 xt rpi5 dom0 wip

GrygiriiS opened the pull request:
First iteration of enabling RPI 5 in Zephyr and adding Dom0 functionality

Lahyllas committed:
drivers: gpio: add brcmstb gpio driver

Add GPIO driver for brcmstb, required by Raspberry Pi 5.

Signed-off-by: Junho Lee <junho@tsnlab.com>
(cherry picked from commit 31baddaa51d5b40e6e6394b8be6cb32e1e962f08)
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Lahyllas committed:
soc: brcm: add support for BCM2712

Add support for BCM2712, SoC of Raspberry Pi 5.

Signed-off-by: Junho Lee <junho@tsnlab.com>
(cherry picked from commit 76ec48179403d6fc15774579e837d0277c00dab6)
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Lahyllas committed:
board: raspberrypi: add support for Raspberry Pi 5

Add support for Raspberry Pi 5.
Currently only internal GPIO(gio_aon) is supported.
'Blinky' is the only sample that is working for now.

Signed-off-by: Junho Lee <junho@tsnlab.com>
(cherry picked from commit 73f4102ad852659baa177ff91c62ebdca6a51596)
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

mypark42 committed:
board: raspberrypi: enable serial communication on Raspberry Pi 5

Enable serial communication through UART port between HDMI ports.

Signed-off-by: Myeonghyeon Park <myeonghyeon@tsnlab.com>
Signed-off-by: Junho Lee <junho@tsnlab.com>
(cherry picked from commit 63692c1349f767c5c78ad30fb4b024c80cb61329)
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
soc: bcm2712: backport to old hwm

Mainline zephyr is moved to HWM v2 which rearranged *soc* and *boards*
which prevents usage of patches [1] as is.

Hence backport bcm2712 to old HWM.

Not for upstream.

[1] https://github.com/zephyrproject-rtos/zephyr/pull/70538
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
boards: raspberrypi: rpi_5: backport to old hwm

Mainline zephyr is moved to HWM v2 which rearranged *soc* and *boards*
which prevents usage of patches [1] as is.

Hence backport RPI5 board to old HWM.

Not for upstream.

[1] https://github.com/zephyrproject-rtos/zephyr/pull/70538
Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
snippets: xen_dom0: add rpi5 board

Add RPI5 board support.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
drivers: xen: add XEN_EVENTS Kconfig option

The XEN events consume 72K of RAM, but they might be not used.

Add XEN EVENTS Kconfig option so XEN events could be gracefully disabled
if not needed.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
boards: arm64: rpi_5: add xen specific doc

Add Xen specific documentation.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
HACK: boards: arm64: xenvm: reduce hypervisor node memory

The current "hypervisor" node defined to use 16M of MMIO memory, which is
way too much. reduce it to 65K. Otherwise it's required to allocate bug
number of grant frames for domain (max_grant_frames).

As suggested by Dmytro Firsov.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
HACK: TMP: xen: public: domctl: tmp adjust version

Tmp adjust XEN_DOMCTL_INTERFACE_VERSION to start with XEN 4.19.

See commit bdb1184d4f6bf4e0121fda34a6c1cb51fe270e7d ("domctl: bump
interface version"), there were no Xen interface struct changes.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

Commit:
boards: arm64: xenvm: defconfig: disable XEN_REGIONS

The CONFIG_XEN_REGIONS is enabled by default, but second memory range is
not defined for "hypervisor" node which causes build failure.

Hence, disable CONFIG_XEN_REGIONS for xenvm platform.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

GrygiriiS commented:
@oleksiimoisieiev @dsemenets @xakep-amatop @Deedone

oleksiimoisieiev commented:
lgtm 
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

GrygiriiS commented:
Some checks failed on RPI5 initial patches as those patches ported as is from Zephyr ML. Those patches will need to be updated later any way when (if) they'll be merged in Zephyr ML. I have no intention to fix those CI warnings now.

UPD: fixed

firscity commented on boards/arm64/xenvm/xenvm_defconfig:
```diff
@@ -18,3 +18,4 @@ CONFIG_LOG_MODE_MINIMAL=n
 CONFIG_USERSPACE=n
 
 CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME=y
+CONFIG_XEN_REGIONS=n
```
It is better to add proper extended region, rather than disabling it completely. They are generated for every domain, you just need to add proper value to device tree. Also, such changes should not be located in board defconfig, it is better to move it to some test sample config you are using for building.

GrygiriiS replied:
Help please - what should be the value for extended region? 

Right now the xenvm build is broken due to ext regions issue, so it has to be fixed within "xenvm" platform instead of adding a w/a fix in each sample/test.

firscity replied:
Oh, I double-checked this config and now I see what is the problem. Okay, let this as is.
We just did not use our Zephyr for building xenvm (since it is not supported/used in any product), only mainline. Will be removed altogether with proper fix.

GrygiriiS replied:
fixed

firscity commented on boards/arm64/rpi_5/Kconfig.board:
```diff
@@ -4,3 +4,7 @@
 	bool "Broadcom BCM2712"
 	depends on SOC_BCM2712
+
+config BOARD_RPI_5_XEN_DOM0
```
This whole commit is something improper. The basic idea of snippets - is extending or overwritting original board behavior: you use RPI5 and want to use Xen dom0 on it? Just apply `xen_dom0` overlay, no other changes or build targets should be needed. All Xen related actions should be done inside a snippet.

GrygiriiS replied:
I simply like this approach more as it allows to keep all platform data in one place, so added it as alternative to snippet. Np. will drop.

firscity replied:
I am looking on all of these patches from upstream perspective, since I am pretty sure that we will need to upstream it. Unfortunately for you, such approach is not upstreamable.

GrygiriiS replied:
removed

firscity commented on submanifests/zephyr-xenlib.yaml:
```diff
@@ -0,0 +1,9 @@
+#
+# Enable zephyr-xenlib support
+
+manifest:
```
Please, provide proper commit message, I am out of context what is happening here and why we need this inside Zephyr

GrygiriiS replied:
It adds  zephyr-xenlib to this particular project. Not sure what to add here :(

https://docs.zephyrproject.org/latest/develop/modules.html#integrate-modules-in-zephyr-build-system

firscity replied:
You should describe not only what you are adding/doing, but also why and how you want to use it here + same point about upstream perspective.

GrygiriiS replied:
dropped

firscity commented on include/zephyr/xen/public/domctl.h:
```diff
@@ -20,7 +20,7 @@
 #include "memory.h"
 
-#define XEN_DOMCTL_INTERFACE_VERSION 0x00000015
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000016
```
Is it actually enough and no changes were made in these structures below? I have some doubts about it

GrygiriiS replied:
It seems enough for this particular exercise. Right change expected to come from https://github.com/xen-troops/zephyr/pull/92 (or similar)

GrygiriiS replied:
FYI. changes in code not in structs
```
commit bdb1184d4f6bf4e0121fda34a6c1cb51fe270e7d
Author: Jan Beulich <jbeulich@suse.com>
Date:   Thu May 11 13:13:55 2023 +0200

    domctl: bump interface version
    
    The change to XEN_DOMCTL_getdomaininfo was a binary incompatible one,
    and the interface version wasn't bumped yet during the 4.18 release
    cycle.
    
    Fixes: 31c655497461 ("domctl: Modify XEN_DOMCTL_getdomaininfo to fail if domid is not found")
    Signed-off-by: Jan Beulich <jbeulich@suse.com>
    Acked-by: Andrew Cooper <andrew.cooper3@citrix.com>

```

firscity commented on boards/arm64/rpi_5/Kconfig.defconfig:
```diff
@@ -17,3 +17,10 @@ config XEN_DOM0
 
 endif # BOARD_RPI_5_XEN_DOM0
+
+if BOARD_RPI_5_XEN_DOMD
```
This is also not okay to made separate board for this. You may use custom sample/application, please check other applications and how they override device tree and Kconfigs.

GrygiriiS replied:
Huh. Sorry, but it will stay here for RPI5 experiments and exercise demonstrations as I can't even imagine how to resolve DT dependencies between clean RPI5, rpi5 dom0, rpi5 domd (uses at least one hw IP) and xenvm. Lets concider samples/basic/blinky:
 - adding rpi_5.overlay for clean RPI5 will work and might work with dom0 snippet, but will not work for "domD".
 - adding RPI5 specific xenvm.overlay will enable support for RPI5 DomD, bit it can't be done for obvious reasons as xenvm is generic.

GrygiriiS replied:
removed and replaced with snippet

firscity reviewed (commented):
Please, fix checkpatch warnings.

Also commits
1. `dts: arm64: broadcom: bcm2712: add psci and compatible` - add proper explanation for this
2. `drivers: xen: add XEN_EVENTS Kconfig option`
`but thay might be not used.` - > `but they might be not used.` 
3. `boards: arm64: xenvm: reduce hypervisor node memory` - please, mark this as a "WA" in commit topic, I will provide a proper solution soon.

Commit:
snippets: add rpi_5_xen_domd to demo hw passing to xen domain

Add rpi_5_xen_domd snippet to demonstrate enabling RPI 5 HW in Xen domain.
Only GPIO LED is supported.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>

GrygiriiS commented:
RPI5 commits went upstream, so cherry-picked and noted.
Other comments fixed.

firscity commented on drivers/xen/Kconfig:
```diff
@@ -29,6 +29,12 @@ config XEN_GRANT_TABLE_INIT_PRIORITY
 	depends on XEN_GRANT_TABLE
 	default 50
 
+config XEN_EVENTS
```
It will be good to take it to `zephyr-v3.6.0-xt` and upstream this change in separate PRs

GrygiriiS replied:
Will note is for integration. and Once we done with basic functionality - I'll revise commits and send those which are generic for integration.

firscity commented on boards/arm64/xenvm/xenvm.dts:
```diff
@@ -69,7 +69,7 @@
 	hypervisor: hypervisor@38000000 {
 		compatible = "xen,xen";
-		reg = <0x00 0x38000000 0x00 0x1000000>;
+		reg = <0x00 0x38000000 0x00 0x10000>;
```
Should be removed after #96 will be merged

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
+========
+
+The Raspberry Pi 5 platform can be used to run as Xen Zephyr Dom0. For such purposes the `xen_dom0`
+can be used.
```
probably you forgot `snippet`

GrygiriiS replied:
fixed

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
+
+   west build -b rpi_5 -p always -S xen_dom0 samples/hello_world
+
+It is expected to be used with special application performing Xen0 functions.
```
`Xen0` -> `Xen Domain-0/Dom0`

GrygiriiS replied:
fixed

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
+============================
+
+The Raspberry Pi 5 platform can be used to run as Xen Zephyr DomU with RPI 5 HW support.
+For such purposes the `rpi_5_xen_domd` can be used.
```
`snippet`

GrygiriiS replied:
fixed

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
+The Raspberry Pi 5 platform can be used to run as Xen Zephyr DomU with RPI 5 HW support.
+For such purposes the `rpi_5_xen_domd` can be used.
+
+Run below command as an example of RPI 5 Zephyr build as Dom0:
```
`Run below command` -> `Run the command below`

`example of RPI 5 Zephyr build as Dom0` - probably copy-paste issue

GrygiriiS replied:
fixed

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
 .. _Raspberry Pi Debug Probe:
    https://www.raspberrypi.com/products/debug-probe/
+
+XEN Dom0
```
Also, I think it is better to make these changes as a part of corresponding commits, which added these functionality (it will be easier to re-arrange it if we will make e.g. DomD snippet as a separate project).

GrygiriiS replied:
I'll keep it in mind.

firscity commented on boards/arm64/rpi_5/doc/index.rst:
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
+   because normaly Xan will update DT for the target Kernel, but this is not possible in case
+   of Zephyr. See comments in `rpi_5_xen_dom0.dts`.
+
+XEN DomU with HW passthrough
```
You used "DomD naming" in snippet, so I think it should be the same here.

GrygiriiS replied:
fixed

Deedone commented on drivers/gpio/gpio_brcmstb.c:
```diff
@@ -0,0 +1,141 @@
+
+static int gpio_brcmstb_pin_configure(const struct device *port, gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct gpio_brcmstb_data *data = port->data;
```
Null check?

GrygiriiS replied:
First it's not my patch and it's explicitly marked as back-ported (with "cherry picked from commit 31baddaa51d5b40e6e6394b8be6cb32e1e962f08", so probably right place for comments is in Zephyr mainline https://github.com/zephyrproject-rtos/zephyr/pull/70538.

Second, those functions are not called directly, but through the gpio APIs and null is impossible here.

Deedone commented on drivers/gpio/gpio_brcmstb.c:
```diff
@@ -0,0 +1,141 @@
+
+static int gpio_brcmstb_port_get_raw(const struct device *port, gpio_port_value_t *value)
+{
+	struct gpio_brcmstb_data *data = port->data;
```
Same, please check all pointers for NULL value.

GrygiriiS replied:
the same as above

Deedone commented on soc/arm64/bcm2712/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_sources_ifdef(CONFIG_ARM_MMU mmu_regions.c)
+
+set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm64/scripts/linker.ld CACHE INTERNAL "")
```
Could you add some comments describing why is this needed pls?

GrygiriiS replied:
same as above

Deedone commented on snippets/rpi_5_xen_domd/README.rst:
```diff
@@ -0,0 +1,17 @@
+.. _rpi_5_xen_domd:
+
+RPI 5 Xen DomD: snippet for XEN HW domain
```
Do we really need separate snippets for DomD and DomUs? AFAIK there is no real difference between them, so what is the reason for explicitly specifying DomD here?

GrygiriiS replied:
if you look at code ypu'll see DomD based on xenvm DT + passedthrough HW, Dom0 has different Xen specific DT and no HW for now. Those are NOT compatible.
